test/e2e/azure_timesync.go

```diff
@@ -94,3 +101,144 @@ func AzureTimeSyncSpec(ctx context.Context, inputGetter func() AzureTimeSyncSpec
 		return kinderrors.AggregateConcurrent(testFuncs)
 	}, thirty, thirty).Should(Succeed())
 }
+
+const (
+	nsenterWorkloadFile = "workloads/nsenter/daemonset.yaml"
+)
+
+// AzureDaemonsetTimeSyncSpec implements a test that verifies time synchronization is healthy for
+// the nodes in a cluster. It uses a privileged daemonset and nsenter instead of SSH.
+func AzureDaemonsetTimeSyncSpec(ctx context.Context, inputGetter func() AzureTimeSyncSpecInput) {
+	var (
+		specName = "azure-timesync"
+		input    AzureTimeSyncSpecInput
+		sixty    = 60 * time.Second
+		five     = 5 * time.Second
+	)
+
+	input = inputGetter()
+	Expect(input.BootstrapClusterProxy).NotTo(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+	namespace, clusterName := input.Namespace.Name, input.ClusterName
+	workloadCluster := input.BootstrapClusterProxy.GetWorkloadCluster(ctx, namespace, clusterName)
+	kubeclient := workloadCluster.GetClient()
+	clientset := workloadCluster.GetClientSet()
+	config := workloadCluster.GetRESTConfig()
+	var nsenterDs unstructured.Unstructured
+
+	yamlData, err := ioutil.ReadFile(nsenterWorkloadFile)
+	Expect(err).NotTo(HaveOccurred(), "failed to unmarshal nsenter daemonset from yaml")
+
+	var jsonData []byte
+	jsonData, err = yaml.YAMLToJSON(yamlData)
+	Expect(err).NotTo(HaveOccurred(), "failed to convert nsenter yaml to json")
+
+	err = nsenterDs.UnmarshalJSON(jsonData)
+	Expect(err).NotTo(HaveOccurred(), "failed to unmarshal nsenter json bytes to unstructured")
+
+	nsenterDs.SetNamespace("default")
+
+	err = kubeclient.Create(ctx, &nsenterDs)
+	Expect(err).NotTo(HaveOccurred(), "failed to create daemonset for time sync check")
+
+	Eventually(func() error {
+		var nodes corev1.NodeList
+		if err := kubeclient.List(ctx, &nodes); err != nil {
```

Comment:
do we need to check this every time? Can we move this outside of eventually?

Reply:
there are two separate eventually loops. this one just waits for the DS pods to come up on all nodes. the other one tries to exec them until it succeeds.

I wrote two loops so we can wait for all nodes/pods read, list out the full set of pods, and then match + retry the exec until success or failure, exactly *without* repeating this inside the full loop more than necessary.

Reply:
IIUC, the nodes should exist by the time it reaches here. So, I thought we could determine the `desired` count before the first eventually loop.

Reply:
took another look in capi, you are right, I can just check the pod counts/readiness here I think 👍

Reply:
think that's better? it was an easy fix actually, since we still need to loop on pod readiness, I think?

Reply:
Yeah, this was my expectation as well. Thanks for fixing it!